lib/handler/gzip.c

```diff
@@ -0,0 +1,184 @@
+        out_size += inbufs[i].len;
+
+    outbuf_index = self->outbufs.size++;
+    self->outbufs.entries[outbuf_index].base = h2o_mem_alloc_pool(&req->pool, out_size);
+    /* set out buffer for output data from zlib */
+    self->zstream.next_out = (unsigned char *) self->outbufs.entries[outbuf_index].base;
+    self->zstream.avail_out = out_size;
+
+    for (i = 0; i != inbufcnt; ++i) {
+        int ret, flush;
+
+        self->zstream.next_in = (unsigned char*) inbufs[i].base;
+        self->zstream.avail_in = inbufs[i].len;
+        
+        if (i == inbufcnt - 1) {
+            flush = is_final ? Z_FINISH : Z_SYNC_FLUSH;
+        } else {
+            flush = Z_NO_FLUSH;
+        }
+        ret = deflate(&self->zstream, flush);
+        if (ret != Z_OK && ret != Z_STREAM_END) {
+            /* Should log and a proper way */
+            assert(0);
+        }
+
+    }
+    self->outbufs.entries[outbuf_index].len = out_size - self->zstream.avail_out;
+    h2o_ostream_send_next(&self->super, req, &self->outbufs.entries[outbuf_index], 1, is_final);
+}
+
+static void on_setup_ostream(h2o_filter_t *self, h2o_req_t *req, h2o_ostream_t **slot)
+{
+    gzip_encoder_t *encoder;
+
+    /* do nothing if content-length is unknown */
+    if (req->res.content_length == SIZE_MAX)
+        goto Next;
+    /* RFC 2616 4.4 states that the following status codes (and response to a HEAD method) should not include message body */
+    if ((100 <= req->res.status && req->res.status <= 199) || req->res.status == 204 || req->res.status == 304)
+        goto Next;
+    else if (h2o_memis(req->input.method.base, req->input.method.len, H2O_STRLIT("HEAD")))
+        goto Next;
+    /* we cannot handle certain responses (like 101 switching protocols) */
+    if (req->res.status != 200) {
+        req->http1_is_persistent = 0;
+        goto Next;
+    }
+    /* skip if transfer-encoding header is being set (for now) */
+    if (h2o_find_header(&req->res.headers, H2O_TOKEN_TRANSFER_ENCODING, -1) != -1)
+        goto Next;
+    /* skip if content-encoding header is being set */
+    if (h2o_find_header(&req->res.headers, H2O_TOKEN_CONTENT_ENCODING, -1) != -1)
+        goto Next;
+    /* 
+     * NOTE: I think there should be some test on MIME type, since we have no need to enable gzip on
+     *       those multimedia files or something like that.
+     */
+
+    /* set content-encoding header */
+    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_ENCODING, H2O_STRLIT("gzip"));
+    /* 
+     * For now, I need to change content_length to SIZE_MAX so I can enable chunked filter.
+     * I need chunked filter because I cannot know exactly how many times the "send_gzip()"
+     * function above will called at this point.
+     */
+    req->res.content_length = SIZE_MAX;
+
+    /* setup filter */
+    encoder = (void *)h2o_add_ostream(req, sizeof(gzip_encoder_t), slot);
+    encoder->super.do_send = send_gzip;
+    slot = &encoder->super.next;
+    
+    /*
```

Comment:
Well it seems that `deflateInit2` is still needed. When I was learning the usage of zlib (well, still not so familiar though), I learned that windowbits should be OR with 16 so zlib will generate gzip header & trailer instead of zlib ones.